rofl-appd/src/routes/tx.rs

```diff
@@ -98,6 +121,68 @@ fn parse_u256_string(value: String) -> Result<evm::types::U256, String> {
         .map_err(|_| "transaction value string is not a valid unsigned integer".to_string())
 }
 
+fn parse_u64_string(value: String, field: &'static str) -> Result<u64, String> {
+    let trimmed = value.trim();
+    if trimmed.is_empty() {
+        return Err(format!("{field} string must not be empty"));
+    }
+    let (radix, digits) = match trimmed
+        .strip_prefix("0x")
+        .or_else(|| trimmed.strip_prefix("0X"))
+    {
+        Some(rest) => (16, rest),
+        None => (10, trimmed),
+    };
+    if digits.is_empty() {
+        return Err(format!("{field} string must contain digits"));
+    }
+    u64::from_str_radix(digits, radix)
+        .map_err(|_| format!("{field} is not a valid unsigned 64-bit integer"))
+}
+
+/// Convert a hex character to its nibble value (0-15).
+#[inline]
+fn hex_nibble(c: u8) -> Option<u8> {
+    match c {
+        b'0'..=b'9' => Some(c - b'0'),
+        b'a'..=b'f' => Some(c - b'a' + 10),
+        b'A'..=b'F' => Some(c - b'A' + 10),
+        _ => None,
+    }
+}
+
+fn deserialize_hex_bytes<'de, D>(deserializer: D) -> Result<Vec<u8>, D::Error>
+where
+    D: serde::Deserializer<'de>,
+{
+    let s = String::deserialize(deserializer)?;
+    let trimmed = s.trim();
+    let without_prefix = trimmed
+        .strip_prefix("0x")
+        .or_else(|| trimmed.strip_prefix("0X"))
+        .unwrap_or(trimmed);
+
+    if without_prefix.is_empty() {
+        // Treat empty (or just "0x") as empty byte sequence.
+        return Ok(Vec::new());
+    }
+    if without_prefix.len() % 2 != 0 {
+        return Err(serde::de::Error::custom("hex string must have even length"));
+    }
+
+    let bytes_slice = without_prefix.as_bytes();
+    let mut bytes = Vec::with_capacity(bytes_slice.len() / 2);
+    for chunk in bytes_slice.chunks_exact(2) {
+        let hi =
+            hex_nibble(chunk[0]).ok_or_else(|| serde::de::Error::custom("invalid hex string"))?;
+        let lo =
+            hex_nibble(chunk[1]).ok_or_else(|| serde::de::Error::custom("invalid hex string"))?;
+        bytes.push((hi << 4) | lo);
+    }
+
+    Ok(bytes)
```

Comment:
I'd rather trim 0x and then use serde's or hex crate to decode to reduce the amount of code.